Generic/UselessOverridingMethod: small improvements to the sniff code

rodrigoprimo opened the pull request:
## Description

This PR implements the following improvements to the Generic.CodeAnalysis.UselessOverridingMethod sniff, which were suggested by @jrfnl in [this comment](https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/250#pullrequestreview-1897345369) in another PR:

> 3. The "Skip for ... other method." block is now covered via the `differentParentMethod()` code snippet.
>     I can see a bug in the condition though - function/method names in PHP are case-insensitive.
>     There is currently no test where the function declaration name is not in the same case as the name used in the function call.
>     Such a test should be added and the code should be flagged as potentially useless override, which would need a small fix in the condition of that code block.
> 4. I'm missing a test where the function call has an open parenthesis, but no close parenthesis.
>     Adding this test would highlight that some extra defensive coding would be helpful.
>     The "Skip for invalid code." block, which checks for the open parenthesis, should probably also check for a matching close parenthesis via `isset($tokens[$next]['parenthesis_closer']) !== false`.
>     The redundant function call on line 118 - `count($tokens)` - can then be also removed and the `$count` in the `for` condition replaced. (That function call was redundant anyhow and should have used `$phpcsFile->numTokens` instead)

> 9. Looking critically, the `$next === false` condition seems redundant in nearly all conditions (and there are no tests conceivable which could ever hit the condition).
>     If the function declaration would not have a scope closer, the scope opener would not be set, so `$next` can never be `false` as there will always be a close curly after whatever code we're looking at.
>     Having said that, if the `$next === false` bits would be removed, it might be prudent to update the "Skip function without body." block to also verify there is a scope closer.
> 10. I'm missing a test with non-OO function declaration. Use of the `parent` keyword in the body of such a function would be invalid no matter what, however, for the purposes of this sniff, I think it would be more appropriate to ignore such functions as they are, for sure, not a "method override".


~~Item number 10 was implemented in two commits. One for removing the redundant `$next === false` condition and another to handle non-OO function declarations.~~ As for handling non-OO function declarations, I changed the sniff code to bail early when the T_FUNCTION does not correspond to a class or anonymous class method. I used `$phpcsFile->hasCondition()` for that. This is the first time that I'm using this method, please let me know if you see a problem with this approach or if there is something that I missed. My assumption is that functions and also trait, interface, and enum methods cannot have a parent so the sniff can bail early in those cases.

Please let me know if you prefer that I split all those fixes into multiple PRs.

## Suggested changelog entry

Generic/UselessOverridingMethod: account for method and parent methods with the same name but different case

I'm unsure if the other changes should be mentioned in the changelog or not. Maybe a second entry referencing "minor performance improvements" (which is most of the other changes but not all other changes)?

## Related issues/external references

Discussed in https://github.com/PHPCSStandards/PHP_CodeSniffer/pull/250#pullrequestreview-1897345369


## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement


## PR checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] [Required for new sniffs] I have added XML documentation for the sniff.

jrfnl commented:
> I changed the sniff code to bail early when the T_FUNCTION does not correspond to a class or anonymous class method. I used $phpcsFile->hasCondition() for that. This is the first time that I'm using this method, please let me know if you see a problem with this approach or if there is something that I missed. 

@rodrigoprimo I haven't looked at the code yet, but thought I'd respond to this comment straight away.

Have you considered nested function declarations ? i.e. https://3v4l.org/f34jJ

> My assumption is that functions and also trait, interface, and enum methods cannot have a parent so the sniff can bail early in those cases.

* Interfaces: correct, not because they can't have a parent (they can, interfaces can extend each other and the class implementing the interface could have a parent class), but because methods in interfaces cannot contain code (and the sniff only looks at the body of methods, not at type declarations).
* Enums: correct, those cannot be extended, so cannot have a parent.
* Trait methods can have a "parent" and should be examined as they are basically "copied in" to the class _using_ the trait and that class can have a parent class.
    It may not be good practice to have assumptions like that in a trait, but that's not the concern of this sniff.

rodrigoprimo commented:
Thanks for your questions, @jrfnl.

> Have you considered nested function declarations ? i.e. https://3v4l.org/f34jJ

Good point. I haven't considered nested functions when working on this.

I did a few tests and modified the original commit to also bail early when it finds a nested function, but now I'm worried that the condition that I introduced might caught more than just nested functions, and I might be introducing a bug. Is there another way to differentiate methods and functions? Maybe the risks of the change that I'm introducing are not worth the benefits? I am happy to follow your advice on this as you have way more experience than I do.

> Trait methods can have a "parent" and should be examined as they are basically "copied in" to the class using the trait and that class can have a parent class.
> It may not be good practice to have assumptions like that in a trait, but that's not the concern of this sniff.

Sorry, my original message could have been more clear. What I meant to say is that a method in an interface, enum, or trait cannot call a parent method as far as I could test. That being said, I can see that I made a mistake when testing traits and that a trait method can actually call a parent method. Thanks for pointing that out.

I have updated the original commit to not bail early for traits, and I included some tests covering trait methods.

rodrigoprimo commented:
@jrfnl, I made some changes to the code that checks if a given function should be checked by the sniff or not after our conversation yesterday, and now this PR is ready for review. Thanks again for your help.

jrfnl commented on src/Standards/Generic/Sniffs/CodeAnalysis/UselessOverridingMethodSniff.php:
```diff
@@ -93,30 +101,29 @@ public function process(File $phpcsFile, $stackPtr)
 
         // Skip for invalid code or other method.
-        if ($next === false || $tokens[$next]['content'] !== $methodName) {
+        if (strcasecmp($tokens[$next]['content'], $methodName) !== 0) {
```
Kudos on using `strcasecmp()` here as "case-insensitive" label names in PHP are only case-insensitive for the ASCII letters and `strcasecmp()` handles this exactly right.

I suspect there are quite a few places in the code base where a `strtolower($name) === $name` type comparison is used instead of `strcasecmp()`. I've opened issue #372 with a task list to review all of those.

rodrigoprimo replied:
Thanks for opening the new issue. I will try to help with that as well in the near future.

jrfnl replied:
Was tempted to just go and fix them, but figured it would be better to open the ticket to allow for (new) contributors to have some more starting points. ;-)

jrfnl commented on src/Standards/Generic/Sniffs/CodeAnalysis/UselessOverridingMethodSniff.php:
```diff
@@ -93,30 +101,29 @@ public function process(File $phpcsFile, $stackPtr)
         }
 
-        // Find next non empty token index, should be the function name.
+        // Find next non empty token index, should be the parent method name.
```
```suggestion
        // Find next non empty token index, should be the method name.
```

rodrigoprimo replied:
I'm not sure if I understand this change. This call to `findNext()` returns the token that corresponds to the parent method name instead of the method name, no?

Using the following code from the test as an example:

```
public function differentParentMethod($a, $b) {
    return parent::anotherMethod($a, $b);
}
```

`findNext()` returns the token for `anotherMethod` and not the token for `differentParentMethod()`.

jrfnl replied:
> I'm not sure if I understand this change. This call to `findNext()` returns the token that corresponds to the parent method name instead of the method name, no?

Well, yes and no. It's the method name _being called_, but still a method name all the same and whether the method exists on the parent is something this sniff will not have any awareness of.

Would "should be the name of the method being called." work for you and make things clearer ?

rodrigoprimo replied:
Oh, when writing the code comment and in my previous comment here, I completely missed that the method could be calling any method from the parent class. Now, your original suggestion sounds good to me, but the last one sounds a bit better, so I went with it.

jrfnl commented on src/Standards/Generic/Tests/CodeAnalysis/UselessOverridingMethodUnitTest.1.inc:
```diff
@@ -64,6 +64,17 @@ class Bar {
         return parent::sameNumberDifferentParameters($this->prop[$a], $this->{$b});
     }
+
+    public function differentCase() {
+        return parent::DIFFERENTcase();
+    }
```
This also needs a test to ensure that method names using non-ASCII characters in different case do not compare as the same name.

Your change should already handle this correctly, but it should still be safeguarded by a test.

rodrigoprimo replied:
@jrfnl, I'm inclined to add this new test right below the other case-related test, but this means updating the line of a few tests in `UselessOverridingMethodUnitTest.php`, which will make reviewing the commit a bit harder, and I remember you already asked me once to add new tests to the end of the file. So I wanted to check with you before proceeding as, in this case, fewer lines will be updated and are lines that did not exist before this PR. Maybe I'm obsessing too much about keeping related tests together in the test file?

jrfnl replied:
@rodrigoprimo You'd only be updating the line numbers of tests being added in this PR, not of pre-existing tests, so all good, though it would be nice if the commits could stay atomic, i.e. the new test would be added to commit 1 and the other commits would be rebased on top of the adjusted first commit and get the line number update in their proper commit.

jrfnl replied:
Note: feel free to make the update in a separate commit though, I can always fix that up before I merge the PR.

rodrigoprimo replied:
I'm happy either way. Whatever makes reviewing and merging the PR easier for you. In this particular case, I added the new test to the original commit to fix the conflicts with the subsequent commits.

jrfnl replied:
@rodrigoprimo I'm looking at the new test now, but it is not actually testing what should be tested...

> method names using non-ASCII characters in a different case do not compare as the same name.

```php
    public function differentCaseWithNonAnsiiCharacters() {
        return parent::DIFFERENTcaseWithNonAnsiiCharáctêrs();
    }
```

👆🏻 This test has different characters (ASCII characters replaced by non-ASCII), which makes this a different method name anyway, which would never be flagged.

```php
    public function Déjàvü() {
        return parent::déJàVü(); // This should be flagged, only ASCII chars have changed case.
    }

    public function Déjàvü() {
        return parent::DÉJÀVÜ(); // This should not be flagged as non-ASCII chars have changed case, making this a different method name.
    }
```

Does that help ?

rodrigoprimo replied:
Apologies for missing this the first time I wrote the test. It should be fixed now.

jrfnl commented on src/Standards/Generic/Sniffs/CodeAnalysis/UselessOverridingMethodSniff.php:
```diff
@@ -56,7 +56,15 @@ public function process(File $phpcsFile, $stackPtr)
+        $lastCondition = end($conditions);
+
+        // Skip functions that are not a method part of a class, anon class or trait.
+        if (in_array($lastCondition, [T_CLASS, T_ANON_CLASS, T_TRAIT]) === false) {
```
Nitpick: as this is a fixed set of tokens anyway, can this array be declared in a class property and checked against via `isset()` ?

This would also prevent potential bugs due to `in_array()` in the code as-is being used without the `$strict` condition. (let me know if you want me to explain that one in more detail)

rodrigoprimo replied:
> Nitpick: as this is a fixed set of tokens anyway, can this array be declared in a class property and checked against via isset() ?

I gave this a try in 73ff2b1c8ced349faf5b3a59e7dbcb926f5feb88. Is this more or less what you had in mind? I'm not super happy with the name of the class property. Please me know if you have a better suggestion.

> This would also prevent potential bugs due to in_array() in the code as-is being used without the $strict condition. (let me know if you want me to explain that one in more detail)

Good point about the lack of `$strict` when calling `in_array()`. I'm assuming you are talking about the possibility of an in_array() behaving unexpectedly as without `$strict` `in_array()` performs a loose comparison, but please let me know if you have something else in mind.

jrfnl replied:
> I gave this a try in [73ff2b1](https://github.com/PHPCSStandards/PHP_CodeSniffer/commit/73ff2b1c8ced349faf5b3a59e7dbcb926f5feb88). Is this more or less what you had in mind?

Is it 👍🏻 

> I'm not super happy with the name of the class property. Please let me know if you have a better suggestion.

What about `$validOOscopes` or even just `$ooScopes` ? (as well as changing the property description to something describing that, like "OO scopes in which a call to `parent::method()` can exist")

> > This would also prevent potential bugs due to in_array() in the code as-is being used without the $strict condition. (let me know if you want me to explain that one in more detail)
> 
> Good point about the lack of `$strict` when calling `in_array()`. I'm assuming you are talking about the possibility of an in_array() behaving unexpectedly as without `$strict` `in_array()` performs a loose comparison, but please let me know if you have something else in mind.

Correct, in particular as PHP native token constants will always be ints, but polyfilled tokens and PHPCS native tokens (like `T_ANON_CLASS`) will always be strings and the strings would be juggled to an int without the `$strict` parameter being `true`.

rodrigoprimo replied:
> What about $validOOscopes or even just $ooScopes ? (as well as changing the property description to something describing that, like "OO scopes in which a call to parent::method() can exist")

Thanks! I went with `$validOOScopes` to hint that the property is not about all the OO scopes.

jrfnl commented on src/Standards/Generic/Tests/CodeAnalysis/UselessOverridingMethodUnitTest.1.inc:
```diff
@@ -76,4 +87,56 @@ interface InterfaceFoo {
+    }
+}
+
+abstract class AbstractBar extends AbstractFoo {
+    public function uselessMethodInAbstractClass() {
+        parent::uselessMethodInAbstractClass();
+    }
+
+    public function usefulMethodInAbstractClass() {
+        $a = 1;
+        parent::usefulMethodInAbstractClass($a);
+    }
+}
```
What is this testing ? (what isn't already tested)

rodrigoprimo replied:
When adding the condition to bail early for functions that are not part of a class, anonymous class, or traits, I added this test to make sure that we had at least the basic tests covering abstract classes. It is true that abstract classes use the same token as regular classes, but I thought that it was still worth adding these tests just in case to avoid regressions in the future, even if, in this case, they are not very likely to happen.

jrfnl replied:
Abstract classes are treated no differently from normal classes by the sniff, abstract methods are, but this class doesn't even have an abstract method..., so it's not testing anything which isn't already being tested.

rodrigoprimo replied:
Yes, I agree that it is not testing anything which isn't already being tested considering the current implementation of the sniff. I'm happy to remove those two tests if you prefer.

> but this class doesn't even have an abstract method

Only today, I realized that I could have added those two methods to the abstract class that already exists in the file. Does that sound better to you? If not, I can simply remove the tests.

jrfnl replied:
Feel free to move them into the existing abstract class. That would make more sense to me.

rodrigoprimo replied:
Done

jrfnl commented on src/Standards/Generic/Tests/CodeAnalysis/UselessOverridingMethodUnitTest.1.inc:
```diff
@@ -76,4 +87,56 @@ interface InterfaceFoo {
+    parent::shouldBailEarly();
+}
+
+new class extends ParentClass {
```
```suggestion
$anon = new class extends ParentClass {
```

rodrigoprimo replied:
I just added this in a new commit, but I wanted to make sure I understand why you suggested this. Is it to highlight that this test is for anonymous classes or something else that I'm missing?

jrfnl replied:
It's mostly to make it realistic code. Without assigning the class to a variable, there would be no way to use the anonymous class.

jrfnl commented on src/Standards/Generic/Tests/CodeAnalysis/UselessOverridingMethodUnitTest.1.inc:
```diff
@@ -64,6 +64,17 @@ class Bar {
+
+    public function nestedFunctionShouldBailEarly() {
+        function nestedFunctionShouldBailEarly() {
+            // Invalid code needed to ensure an error is NOT triggered and the sniff bails early when handling nested function.
```
☝🏻 These comments aren't really needed. (though don't hurt either, so no need to remove them, just wanted to point it out)

rodrigoprimo replied:
I added this comment as, to me, it is not immediately obvious why calling `parent::nestedFunctionShouldBailEarly()` is needed inside the function, and in the future, others might wonder the same as this is a fatal error.

jrfnl replied:
Well, the test wouldn't be a valid test without the `parent::...()` call as it wouldn't fail if the code causing the sniff to bail early would be removed....

Either way, I'm fine with leaving the comments be

jrfnl commented on src/Standards/Generic/Tests/CodeAnalysis/UselessOverridingMethodUnitTest.1.inc:
```diff
@@ -76,4 +87,56 @@ interface InterfaceFoo {
+};
+
+function foo() {
+    new class extends ParentClass {
```
```suggestion
    $anon = new class extends ParentClass {
```

jrfnl reviewed (changes requested):
Hi @rodrigoprimo, thanks for working on this. Looking good! Just needs some dotting of the i's and crossing of the t's.

P.S.: Sorry, I made a small edit in the PR description as the quoted text didn't line up with the comment in the other PR (which was confusing me).

jrfnl commented:
> Please let me know if you prefer that I split all those fixes into multiple PRs.

@rodrigoprimo No need. The way you've done it now with atomic commits for each change is exactly how I like it and allows for reviewing each change individually 👍🏻

rodrigoprimo commented:
@jrfnl, thanks for all your comments. This PR is ready for another review.

jrfnl commented on src/Standards/Generic/Sniffs/CodeAnalysis/UselessOverridingMethodSniff.php:
```diff
@@ -28,6 +28,18 @@
+     * List of class and trait tokens. Used to detect functions that
+     * are part of a class, anon class or trait.
+     *
+     * @var array
```
```suggestion
     * @var array<int|string, bool> Keys are the token constants, value is irrelevant.
```

rodrigoprimo commented:
This PR is ready for another review.

jrfnl reviewed (approved):
Thanks @rodrigoprimo ! This looks good to me now. Nice clean up and improvements for this sniff.

I will rebase the PR interactively to get back to the original atomic commits and then merge the PR once the build has passed.
